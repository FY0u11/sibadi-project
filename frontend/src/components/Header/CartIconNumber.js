import React from 'react'

const CartIconNumber = ({ number, radius, className }) => {
  return (
    <svg
      className={className}
      height={radius * 2}
      width={radius * 2}
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle cx={radius} cy={radius} r={radius} fill="red" />
      <text
        y={Math.floor(radius * 2 - radius * 0.3)}
        x={radius}
        fontFamily="sans-serif"
        fontSize={radius * 2}
        fontWeight="bold"
        fill="white"
        textAnchor="middle"
      >
        {number}
      </text>
    </svg>
  )
}

export default CartIconNumber
