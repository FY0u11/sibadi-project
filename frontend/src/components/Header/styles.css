.my-page-title {
  display: inline-block;
  margin-left: 20px;
  padding-left: 20px;
  border-left: 1px solid grey;
}

.my-icon {
  margin-left: 1rem;
}

.my-logout-icon {
  display: inline;
  margin-left: 2.5rem;
}

.my-cart-icon-container {
  position: relative;
}

.my-cart-icon {
  position: absolute;
  top: 2px;
}

.my-cart-icon-number {
  position: absolute;
  top: 13px;
  left: 11px;
}
