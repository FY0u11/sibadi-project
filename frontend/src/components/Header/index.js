import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import { useLocation } from 'react-router-dom'

import { setHeaderTitle } from '../../store/components/header/actions'
import { setIsCartShown } from '../../store/components/cart/actions'

import Header from './Header'

const headerTitles = {
  '/main/profile': 'Profile',
  '/main/profile/edit': 'Profile > Edit profile',
  '/main/profile/changePassword': 'Profile > Change password',
  '/main/products': 'Products',
  '/main/delivery': 'Delivery',
  '/main/orders': 'Orders',
  '/main/requests': 'Requests',
  '/authorization/login': 'Login',
  '/authorization/register': 'Register'
}

const HeaderContainer = ({ setHeaderTitle, ...props }) => {
  const currentLocation = useLocation()

  useEffect(() => {
    setHeaderTitle(headerTitles[currentLocation.pathname])
  }, [currentLocation])

  return <Header {...props} />
}

const putStateToProps = state => ({
  user: state.root.user,
  isCartShown: state.cart.isCartShown,
  cartProducts: state.cart.cartProducts,
  headerTitle: state.header.headerTitle
})

const putActionsToProps = { setHeaderTitle, setIsCartShown }

export default connect(putStateToProps, putActionsToProps)(HeaderContainer)
