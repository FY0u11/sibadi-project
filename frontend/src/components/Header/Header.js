import React from 'react'
import { Link } from 'react-router-dom'
import { IoMdCart } from 'react-icons/io'
import { FaSignOutAlt, FaUser } from 'react-icons/fa'

import './styles.css'
import CartIconNumber from './CartIconNumber'

const Header = ({ user, isCartShown, setIsCartShown, cartProducts, headerTitle }) => {
  return (
    <header className="navbar fixed-top navbar-dark bg-dark">
      <div className="container">
        <div className="navbar-brand">
          <Link to="/">
            Welcome to <strong>Azamon</strong>
          </Link>
          <div className="my-page-title">{headerTitle}</div>
        </div>
        <div className="navbar-brand">
          {user && (
            <div>
              <Link to="/main/profile">
                <strong>{user.username}</strong>
                <FaUser className="my-icon" />
              </Link>
              <Link
                to="#"
                className="my-icon my-cart-icon-container"
                onClick={() => setIsCartShown(!isCartShown)}
                data-testid="toggleCart"
              >
                <IoMdCart className="my-cart-icon" />
                <CartIconNumber
                  className="my-cart-icon-number"
                  number={cartProducts.length}
                  radius="6"
                />
              </Link>
              <div className="my-logout-icon">
                <Link to="/main/logout">
                  <FaSignOutAlt />
                </Link>
              </div>
            </div>
          )}
        </div>
      </div>
    </header>
  )
}

export default Header
