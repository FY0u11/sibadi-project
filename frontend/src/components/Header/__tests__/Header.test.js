import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
import { render, cleanup, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Header from '../Header'

afterEach(cleanup)

describe('Testing Header component', () => {
  it('It should be properly rendered without user property', () => {
    const { asFragment } = render(
      <Router>
        <Header headerTitle="Header" />
      </Router>
    )
    expect(asFragment()).toMatchSnapshot('1')
  })
  it('It should be properly rendered with user property', () => {
    const user = { username: 'John' }
    const { asFragment, getByText } = render(
      <Router>
        <Header headerTitle="Header" user={user} cartProducts={[{}, {}, {}, {}, {}]} />
      </Router>
    )
    expect(asFragment()).toMatchSnapshot('2')
    expect(getByText('John')).toBeInTheDocument()
  })
  it('It should toggle a variable with onClick', () => {
    const user = { username: 'John' }
    let isCartShown = false
    const props = {
      user,
      cartProducts: [{}, {}, {}, {}, {}],
      isCartShown,
      setIsCartShown: bool => {
        isCartShown = bool
      }
    }
    const { getByTestId } = render(
      <Router>
        <Header {...props} />
      </Router>
    )
    expect(isCartShown).toBeFalsy()
    fireEvent.click(getByTestId('toggleCart'))
    expect(isCartShown).toBeTruthy()
  })
})
