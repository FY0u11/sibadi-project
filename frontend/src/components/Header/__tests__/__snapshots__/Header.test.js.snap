// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Header component It should be properly rendered with user property: 2 1`] = `
<DocumentFragment>
  <header
    class="navbar fixed-top navbar-dark bg-dark"
  >
    <div
      class="container"
    >
      <div
        class="navbar-brand"
      >
        <a
          href="/"
        >
          Welcome to 
          <strong>
            Azamon
          </strong>
        </a>
        <div
          class="my-page-title"
        >
          Header
        </div>
      </div>
      <div
        class="navbar-brand"
      >
        <div>
          <a
            href="/main/profile"
          >
            <strong>
              John
            </strong>
            <svg
              class="my-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
              />
            </svg>
          </a>
          <a
            class="my-icon my-cart-icon-container"
            data-testid="toggleCart"
            href="/"
          >
            <svg
              class="my-cart-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M169.6 377.6c-22.882 0-41.6 18.718-41.6 41.601 0 22.882 18.718 41.6 41.6 41.6s41.601-18.718 41.601-41.6c-.001-22.884-18.72-41.601-41.601-41.601zM48 51.2v41.6h41.6l74.883 151.682-31.308 50.954c-3.118 5.2-5.2 12.482-5.2 19.765 0 27.85 19.025 41.6 44.825 41.6H416v-40H177.893c-3.118 0-5.2-2.082-5.2-5.2 0-1.036 2.207-5.2 2.207-5.2l20.782-32.8h154.954c15.601 0 29.128-8.317 36.4-21.836l74.882-128.8c1.237-2.461 2.082-6.246 2.082-10.399 0-11.446-9.364-19.765-20.8-19.765H135.364L115.6 51.2H48zm326.399 326.4c-22.882 0-41.6 18.718-41.6 41.601 0 22.882 18.718 41.6 41.6 41.6S416 442.082 416 419.2c0-22.883-18.719-41.6-41.601-41.6z"
              />
            </svg>
            <svg
              class="my-cart-icon-number"
              height="12"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="6"
                cy="6"
                fill="red"
                r="6"
              />
              <text
                fill="white"
                font-family="sans-serif"
                font-size="12"
                font-weight="bold"
                text-anchor="middle"
                x="6"
                y="10"
              >
                5
              </text>
            </svg>
          </a>
          <div
            class="my-logout-icon"
          >
            <a
              href="/main/logout"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M497 273L329 441c-15 15-41 4.5-41-17v-96H152c-13.3 0-24-10.7-24-24v-96c0-13.3 10.7-24 24-24h136V88c0-21.4 25.9-32 41-17l168 168c9.3 9.4 9.3 24.6 0 34zM192 436v-40c0-6.6-5.4-12-12-12H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h84c6.6 0 12-5.4 12-12V76c0-6.6-5.4-12-12-12H96c-53 0-96 43-96 96v192c0 53 43 96 96 96h84c6.6 0 12-5.4 12-12z"
                />
              </svg>
            </a>
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Testing Header component It should be properly rendered without user property: 1 1`] = `
<DocumentFragment>
  <header
    class="navbar fixed-top navbar-dark bg-dark"
  >
    <div
      class="container"
    >
      <div
        class="navbar-brand"
      >
        <a
          href="/"
        >
          Welcome to 
          <strong>
            Azamon
          </strong>
        </a>
        <div
          class="my-page-title"
        >
          Header
        </div>
      </div>
      <div
        class="navbar-brand"
      />
    </div>
  </header>
</DocumentFragment>
`;
